### ISSUE TITLE
Move KWS preprocessing layers and fail to save model

### ISSUE THREAD

**Issue by tu1258** [google-research/google-research#408]
I put some work to minimize the code. What I exactly do is to move the sublayers inside SpeechFeature layer to model.py(let's say tc_resnet.py for instance). But now I fail to save model.(It can be trained without any bug.) This is not a bug, it's just my dumb question instead but maybe @rybakov know what I should do? I am sorry to bother you again. I've done so many changes but the the code structure doesn't change much.

### To make it clear, this is what I do. I think you would understand instantly 😄
```
  # net = speech_features.SpeechFeatures(
  #       speech_features.SpeechFeatures.get_params(flags),
  #       name='feature_layer')(
  #         net)
  net = dataframe.DataFrame(
        inference_batch_size=1,
        frame_size=frame_size,
        frame_step=frame_stride)(
        net)
  net = windowing.Windowing(
        window_size=frame_size, 
        window_type=flags.window_type)(
        net)
  net = magnitude_rdft_mel.MagnitudeRDFTmel(
        num_mel_bins=flags.mel_num_bins,
        lower_edge_hertz=flags.mel_lower_edge_hertz,
        upper_edge_hertz=flags.mel_upper_edge_hertz,
        sample_rate=flags.sample_rate,
        mel_non_zero_only=flags.mel_non_zero_only)(
        net)
  net = Lambda(lambda x: tf.math.log(tf.math.maximum(x, flags.log_epsilon)))(net)
  net = dct.DCT(num_features=flags.dct_num_features)(net)
```

The error I got when running this line `model.save(path_model, include_optimizer=False, save_format='tf')`
**Error**
`'Not JSON Serializable:', Namespace(background_frequency=0.8, background_volume=0.1, batch_size=1, bn=0, channels='36, 36, 36, 36, 36, 36, 36', clip_duration_ms=1000, data_dir='./data_test', dct_num_features=30, debug=0, desired_samples=16000, dropout=0.0, eval_step_interval=200, first_kernel='(1,1)', generate=0, groups=1, how_many_training_epochs='1', kernel_size='(3,1)', label_count=12, learning_rate='0.001', log_epsilon=1e-12, lr_schedule='linear', mel_lower_edge_hertz=20.0, mel_non_zero_only=1, mel_num_bins=80, mel_upper_edge_hertz=7000.0, mobile=1, optimizer='adam', preemph=0.0, print=0, qkeras=0, resample=0.15, residual=0, sample_rate=16000, save_step_interval=1000, silence_percentage=10.0, spectrogram_length=49, start_checkpoint='', summaries_dir='./tcn/kws_test/logs/', testing_percentage=10, time_shift_ms=100.0, train=1, train_dir='./tcn/kws_test', unknown_percentage=10.0, validation_percentage=10, volume_resample=0.0, wanted_words='yes,no,up,down,left,right,on,off,stop,go', window_size_ms=40.0, window_size_samples=640, window_stride_ms=20.0, window_stride_samples=320, window_type='hann'))`

**Comment by rybakov** [comment]
This error says that you can not save flags (all parametrs there are belong to global flag). It is not clear why you are getting this error. I created a colab with such a model and it works. I also would suggest to save just model weights: model.save_weights("/tmp/01/"), please find attached example 
[test.zip](https://github.com/google-research/google-research/files/5303135/test.zip)

**Comment by tu1258** [comment]
@rybakov Thanks for your reply! However, I couldn't find the bug in my code even if I could reproduce your results and it seems so weird. If you don't mind to try, I build a repository and [google colab](https://colab.research.google.com/drive/1V9_Whe-8lmiVtxM4Zpe1mBWrBVwyhmnG?usp=sharing) to reproduce my results. All can be done in 10-15 min. The google colab should be a successful example. And if you change the code in `kws_test/train/tcn.py`

from
```
  net = speech_features.SpeechFeatures(
        speech_features.SpeechFeatures.get_params(flags),
        name='feature_layer')(
          net)
  # net = dataframe.DataFrame(
  #       inference_batch_size=1,
  #       frame_size=frame_size,
  #       frame_step=frame_stride)(
  #       net)
  # net = windowing.Windowing(
  #       window_size=frame_size, 
  #       window_type=flags.window_type)(
  #       net)
  # net = magnitude_rdft_mel.MagnitudeRDFTmel(
  #       num_mel_bins=flags.mel_num_bins,
  #       lower_edge_hertz=flags.mel_lower_edge_hertz,
  #       upper_edge_hertz=flags.mel_upper_edge_hertz,
  #       sample_rate=flags.sample_rate,
  #       mel_non_zero_only=flags.mel_non_zero_only)(
  #       net)
  # net = Lambda(lambda x: tf.math.log(tf.math.maximum(x, flags.log_epsilon)))(net)
  # net = dct.DCT(num_features=flags.dct_num_features)(net)
```
to
```
  # net = speech_features.SpeechFeatures(
  #       speech_features.SpeechFeatures.get_params(flags),
  #       name='feature_layer')(
  #         net)
  net = dataframe.DataFrame(
        inference_batch_size=1,
        frame_size=frame_size,
        frame_step=frame_stride)(
        net)
  net = windowing.Windowing(
        window_size=frame_size, 
        window_type=flags.window_type)(
        net)
  net = magnitude_rdft_mel.MagnitudeRDFTmel(
        num_mel_bins=flags.mel_num_bins,
        lower_edge_hertz=flags.mel_lower_edge_hertz,
        upper_edge_hertz=flags.mel_upper_edge_hertz,
        sample_rate=flags.sample_rate,
        mel_non_zero_only=flags.mel_non_zero_only)(
        net)
  net = Lambda(lambda x: tf.math.log(tf.math.maximum(x, flags.log_epsilon)))(net)
  net = dct.DCT(num_features=flags.dct_num_features)(net)
```
It shows the error I mentioned before.

I have done some changes but if those matters, it wouldn't be able to save model successfully when using SpeechFeature layer. I believe the bug is in feature files(`dataframe, windowing, magnitude_rdft_mel, dct`). I've tried with the original feature files, but still can't save model. I am absolutely fine if you don't have time or will to reproduce. Much appreciated!

**Comment by tu1258** [comment]
Thanks for your reply! It turns out that I shouldn't use flags in `net = Lambda(lambda x: tf.math.log(tf.math.maximum(x, flags.log_epsilon)))(net)` in my case. I'll close this issue. Thanks for the help, you are always so kind! ❤️
